ably/export_test.go

```diff
@@ -193,6 +205,49 @@ func (c *Connection) PendingItems() int {
 	return len(c.pending.queue)
 }
 
+type MsgWithAckCallback = msgWithAckCallback
+
+// AckAll empties queue and acks all pending callbacks
+func (c *Connection) AckAll() {
+	c.mtx.Lock()
+	cx := c.pending.Dismiss()
+	c.mtx.Unlock()
+	c.log().Debugf("resending %d messages waiting for ACK/NACK", len(cx))
+	for _, v := range cx {
+		v.onAck(nil)
+	}
+}
+
+func (c *Connection) SetKey(key string) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.key = key
+}
+
+func (c *RealtimePresence) GetMembers() map[string]*PresenceMessage {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.members
```

Comment:
If this map might get mutated after this call returns, then holding the lock just to return the map isn't enough, the map itself needs to be thread safe.

Reply:
actually these methods are only used for `testing`, are not used anywhere else.
All methods defined under `export_test.go` are available only for tests.

Reply:
Ok, but even in the tests we should be concerned about making things thread safe, otherwise we'll get failing tests due to races being detected.

Reply:
updated -> https://github.com/ably/ably-go/pull/621/commits/31b739a380c961ff176671093579c8928e4b9792